<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jon Thompson, Pers lab" />


<title>Single cell data analysis workshop</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">190510-scWorkshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/JonThom/190510-scWorkshop">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Single cell data analysis workshop</h1>
<h4 class="author"><em>Jon Thompson, Pers lab</em></h4>
<h4 class="date"><em>2019-05-10 11:05:12</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-05-10
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>190510-scWorkshop/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.3.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190507code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190507code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190507)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomJonThom190510scWorkshoptree7aa71e440c4c8ce056776b4bf09227aab47a6556targetblank7aa71e4a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/JonThom/190510-scWorkshop/tree/7aa71e440c4c8ce056776b4bf09227aab47a6556" target="_blank">7aa71e4</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomJonThom190510scWorkshoptree7aa71e440c4c8ce056776b4bf09227aab47a6556targetblank7aa71e4a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  data/filtered_gene_bc_matrices/
    Untracked:  data/pbmc3k_filtered_gene_bc_matrices.tar.gz
    Untracked:  output/seu_pbmc_tutorial.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/7aa71e440c4c8ce056776b4bf09227aab47a6556/analysis/analysis.Rmd" target="_blank">7aa71e4</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
wflow_publish(“./analysis/analysis.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/1d5ea1fcda8330d20343ca2222451609a95f5508/docs/analysis.html" target="_blank">1d5ea1f</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/0b4f952b473b5653698f687b2f68ddc63337e035/analysis/analysis.Rmd" target="_blank">0b4f952</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
wflow_publish(“./analysis/analysis.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/966c250a78bade4d0b7c9be1210aff5c53b79e3b/docs/analysis.html" target="_blank">966c250</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/6acaec4b34edf9d57adf1caf023f1835cf43813b/analysis/analysis.Rmd" target="_blank">6acaec4</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
publish
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/1bfc8f18551e969cb657904cd02fb0b99f4b51e4/docs/analysis.html" target="_blank">1bfc8f1</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/080696fe55c4b9534e25233614e1941d2a18f224/analysis/analysis.Rmd" target="_blank">080696f</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
wflow_publish(“./analysis/analysis.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/7c8ff3e04ab986873509ce7ee7ed8bb6da84e31a/docs/analysis.html" target="_blank">7c8ff3e</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/9934dbb7a145a3df0ed4260467d576575342bfb3/analysis/analysis.Rmd" target="_blank">9934dbb</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
<td>
publish
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/11811b9b5d1ce1fa7a53801f4ce4fb3218d9657b/docs/analysis.html" target="_blank">11811b9</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/25d1dfcd4c406305445325d3676756b38fe7a53a/analysis/analysis.Rmd" target="_blank">25d1dfc</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
<td>
working draft
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/ed6f27ad62f4a7b2a3d95b9aea74027f40ba61af/docs/analysis.html" target="_blank">ed6f27a</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/fcf3e9123afdba50085b224644448da495963e5f/docs/analysis.html" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/eb3e7f57fa375853d6e4285410a6eb5db75dfc0c/analysis/analysis.Rmd" target="_blank">eb3e7f5</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
<td>
wflow_publish(files = “./analysis/analysis.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/JonThom/190510-scWorkshop/129883be4b2e481c0ea27a2b56103816d7d3d799/docs/analysis.html" target="_blank">129883b</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/a90671853b796a72239b5d1e72374c3ec7b3162f/analysis/analysis.Rmd" target="_blank">a906718</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-07
</td>
<td>
add first analysis file
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>A lightly adapted and annotated version of Satija et al, <a href="https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html">Guided Clustering Tutorial</a> (2019)</p>
<ul>
<li>Seurat is an open source R package developed by Rahul Satija’s group at New York University, first released spring 2015</li>
<li>Widely used and under active development</li>
<li>Seurat is part of an ecosystem of R packages
<ul>
<li>e.g. Seurat is built on <a href="https://ggplot2.tidyverse.org/">ggplot2</a></li>
<li>specialized tools like <a href="https://www.biorxiv.org/content/10.1101/352484v3">DoubletFinder</a> or <a href="https://www.biorxiv.org/content/10.1101/303727v1">SoupX</a> are built on Seurat</li>
</ul></li>
<li>There are great tutorials available on <a href="https://github.com/satijalab/seurat">Satija group website</a></li>
</ul>
<p>This tutorial covers the standard steps of QC, pre-processing and selecting features (genes), reducing the dimensionality of your dataset, clustering cells, and finding differentially expressed features.</p>
<p>Seurat has many other tools, notably for integrating different datasets, which other presenters will cover today.</p>
<p>Download the script for this tutorial at <a href="https://github.com/JonThom/190510-scWorkshop/blob/master/analysis/analysis.Rmd">github</a></p>
<div id="set-up" class="section level1">
<h1>Set up</h1>
<p>install and load packages</p>
<pre class="r"><code>pkgs_required &lt;- c(&quot;Seurat&quot;, &quot;dplyr&quot;)

pkgs_new &lt;- pkgs_required[!(pkgs_required %in% installed.packages()[, &quot;Package&quot;])]

if (length(pkgs_new)&gt;0) install.packages(pkgs_new)

suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(dplyr))</code></pre>
<p>Set standard options</p>
<pre class="r"><code>options(warn=1, stringsAsFactors = F)</code></pre>
<p>Define constants</p>
<pre class="r"><code># Define random seed for reproducibility
randomSeed = 12345
set.seed(seed=randomSeed)</code></pre>
<div id="load-data-and-initialize-the-seurat-object" class="section level2">
<h2>Load data and initialize the Seurat Object</h2>
<p>For this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">freely available from 10X Genomics</a>. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500 and aligned to the human genome using <code>Cell Ranger</code>.</p>
<div id="reading-in-data" class="section level3">
<h3>Reading in data</h3>
<p>Seurat takes as input a gene (row) * cell (column) matrix of transcript counts.</p>
<p>The <code>Read10X</code> function reads in the output of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger</a> pipeline from 10X, returning a unique molecular identified (UMI) count matrix.</p>
<p>Seurat also provides the function <code>ReadAlevin</code> for loading the binary format matrix produced by the <a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin tool from the Salmon software</a>.</p>
<ul>
<li>Before loading the filtered matrix, you should be confident that previous filtering steps were sound. If using <code>Cell Ranger</code>, inspect the outputs carefully to see if the cut for calling barcodes as cells seems reasonable. If not, consider using the unfiltered matrix and filtering manually e.g. on a minimum number of RNA counts per cell.</li>
</ul>
<pre class="r"><code># Load the PBMC dataset
pbmc.data &lt;- Read10X(data.dir = &quot;/projects/jonatan/applied/190510-scWorkshop/data/filtered_gene_bc_matrices/hg19/&quot;)</code></pre>
</div>
<div id="create-a-seurat-object" class="section level3">
<h3>Create a Seurat object</h3>
<p>We next use the count matrix to create a <code>Seurat</code> object, which serves as a container for data (like the count matrix) and analysis (like PCA, or clustering results).</p>
<pre class="r"><code># Initialize the Seurat object with the raw (non-normalized data).
pbmc &lt;- suppressWarnings({
  CreateSeuratObject(counts = pbmc.data, 
                     project = &quot;pbmc3k&quot;, 
                     min.cells = 3, # set minimum number of cells for genes
                     min.features = 200) # set minimum number of features (genes) for cells
  })
pbmc</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features)</code></pre>
<p>Seurat provides convenience functions to access data in the object:</p>
<pre class="r"><code>GetAssayData(pbmc, slot=&quot;counts&quot;)[0:5,0:3]</code></pre>
<pre><code>5 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
              AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC
AL627309.1                 .              .              .
AP006222.2                 .              .              .
RP11-206L10.2              .              .              .
RP11-206L10.9              .              .              .
LINC00115                  .              .              .</code></pre>
<p>In addition, you can always access any slot in the object using the ‘@’ and ‘$’ operators:</p>
<pre class="r"><code>pbmc@assays$RNA@counts[0:5,0:3]</code></pre>
<pre><code>5 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
              AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC
AL627309.1                 .              .              .
AP006222.2                 .              .              .
RP11-206L10.2              .              .              .
RP11-206L10.9              .              .              .
LINC00115                  .              .              .</code></pre>
<p>For a technical discussion of the Seurat object structure, check out the <a href="https://github.com/satijalab/seurat/wiki">GitHub Wiki</a>.</p>
</div>
</div>
</div>
<div id="qc-and-normalization" class="section level1">
<h1>QC and normalization</h1>
<div id="standard-pre-processing-workflow-filter-cells-based-on-qc-metrics" class="section level2">
<h2>Standard pre-processing workflow: filter cells based on QC metrics</h2>
<p>Seurat allows you to easily explore QC metrics and filter cells based on <em>any</em> user-defined criteria.</p>
<p>A few QC metrics <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/">commonly used</a> by the community include</p>
<ul>
<li>The number of unique genes or molecules detected in each cell.
<ul>
<li>Low-quality cells or empty droplets will often have very few genes</li>
<li>Cell doublets or multiplets may exhibit an aberrantly high gene count</li>
</ul></li>
<li>The percentage of counts that map to the mitochondrial genome. Low-quality / dying cells often exhibit extensive mitochondrial contamination</li>
<li>Depending on the particular experiment, other indicators might include e.g. proportion of counts mapping to the ribosomal genome, in some cases a marker for cell stress.
<ul>
<li>Exercise caution! According to <a href="https://kb.10xgenomics.com/hc/en-us/articles/218169723-What-fraction-of-reads-map-to-ribosomal-proteins-">10x Genomics</a>, for PBMCs approximately 35-40% of reads map to ribosomal transcripts. Mitochrondrial RNA content also varies by celltype and condition.</li>
</ul></li>
</ul>
<pre class="r"><code>mito.genes &lt;- grepl(pattern = &quot;^mt-&quot;, x = rownames(pbmc), ignore.case=T)

mat_counts &lt;- GetAssayData(object = pbmc, assay=&quot;RNA&quot;, slot = &quot;counts&quot;) %&gt;% as.matrix

colSums_tmp &lt;- colSums(x = mat_counts)

pbmc[[&quot;percent.mt&quot;]] = colSums(x = mat_counts[mito.genes,])/colSums_tmp

head(pbmc@meta.data)</code></pre>
<pre><code>               orig.ident nCount_RNA nFeature_RNA  percent.mt
AAACATACAACCAC     pbmc3k       2419          779 0.030177759
AAACATTGAGCTAC     pbmc3k       4903         1352 0.037935958
AAACATTGATCAGC     pbmc3k       3147         1129 0.008897363
AAACCGTGCTTCCG     pbmc3k       2639          960 0.017430845
AAACCGTGTATGCG     pbmc3k        980          521 0.012244898
AAACGCACTGGTAC     pbmc3k       2163          781 0.016643551</code></pre>
<pre class="r"><code># Visualize QC metrics as a violin plot
VlnPlot(pbmc, 
        features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), 
        ncol = 3)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-9-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/1d5ea1fcda8330d20343ca2222451609a95f5508/docs/figure/analysis.Rmd/unnamed-chunk-9-1.png" target="_blank">1d5ea1f</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-9-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/1d5ea1fcda8330d20343ca2222451609a95f5508/docs/figure/analysis.Rmd/unnamed-chunk-10-1.png" target="_blank">1d5ea1f</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-10-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)
# plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)
# CombinePlots(plots = list(plot1, plot2))</code></pre>
<ul>
<li>We filter cells that have unique feature counts over 2,500 (indicating a possible multiplet) or fewer than 200</li>
<li>We filter cells that have &gt;5% mitochondrial counts</li>
</ul>
<pre class="r"><code>pbmc &lt;- subset(pbmc, 
               subset = nFeature_RNA &gt; 200 &amp; 
                 nFeature_RNA &lt; 2500 &amp; 
                 percent.mt &lt; 0.05)

head(pbmc@meta.data)</code></pre>
<pre><code>               orig.ident nCount_RNA nFeature_RNA  percent.mt
AAACATACAACCAC     pbmc3k       2419          779 0.030177759
AAACATTGAGCTAC     pbmc3k       4903         1352 0.037935958
AAACATTGATCAGC     pbmc3k       3147         1129 0.008897363
AAACCGTGCTTCCG     pbmc3k       2639          960 0.017430845
AAACCGTGTATGCG     pbmc3k        980          521 0.012244898
AAACGCACTGGTAC     pbmc3k       2163          781 0.016643551</code></pre>
</div>
<div id="lognormalize-the-raw-counts" class="section level2">
<h2>LogNormalize the raw counts</h2>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data.</p>
<p>By default, we employ a global-scaling normalization method <code>LogNormalize</code> that</p>
<ol style="list-style-type: decimal">
<li>normalizes the feature expression measurements for each cell by the total number of counts within the cell and scales this by a common factor (10,000 by default) corresponding to the expected number of reads in a cell
<ul>
<li>assumption: differences in total number of RNA are due to more varying sampling depth than to biology.</li>
</ul></li>
<li>adds 1 and log-transforms (natural log) the result.
<ul>
<li>assumption: raw counts are highly skewed to the right and log brings them closer to Normality, which is useful for PCA and statistical tests.</li>
</ul></li>
</ol>
<p>Normalized values are stored in <code>pbmc@assays$RNA@data</code>.</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc, 
                      normalization.method = &quot;LogNormalize&quot;, 
                      scale.factor = 10000)

pbmc@assays$RNA@data[10000:10005,0:3]</code></pre>
<pre><code>6 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
        AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC
HDGFRP3       .                     .       .       
ZSCAN2        .                     .       .       
WDR73         .                     .       .       
NMB           .                     .       .       
SEC11A        1.635873              .       1.429744
ZNF592        .                     .       .       </code></pre>
</div>
<div id="identify-highly-variable-features-feature-selection" class="section level2">
<h2>Identify highly variable features (feature selection)</h2>
<p>We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Satija et al and <a href="https://www.nature.com/articles/nmeth.2645">others</a> have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.</p>
<p><em>Problem: gene expression variance tends to correlate highly with the mean expression, leading to a bias towards selecting highly expressed genes</em></p>
<p><strong>Variance Stabilizing Transformation (vst)</strong>:</p>
<ul>
<li>The procedure in Seurat3 is described in detail <a href="https://www.biorxiv.org/content/early/2018/11/02/460147.full.pdf">here</a></li>
<li>The method fits a line to the relationship of log(variance) and log(mean) across all genes using local polynomial regression (loess). Then standardizes each gene’s variance using the observed mean and expected variance (given by the fitted line). To reduce the impact of technical outliers, we clip the standardized values to a maximum value (clip.max).</li>
<li>This method returns the number of genes requested (default 2000)</li>
</ul>
<pre class="r"><code>pbmc &lt;- FindVariableFeatures(pbmc, 
                             selection.method = &quot;vst&quot;, 
                             nfeatures = 2000)</code></pre>
<pre class="r"><code># Identify the 10 most highly variable genes
top10 &lt;- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 &lt;- VariableFeaturePlot(pbmc)
plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)</code></pre>
<pre><code>When using repel, set xnudge and ynudge to 0 for optimal results</code></pre>
<pre class="r"><code>CombinePlots(plots = list(plot1, plot2))</code></pre>
<pre><code>Warning: Transformation introduced infinite values in continuous x-axis

Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-14-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-14-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="scale-and-center-the-data" class="section level2">
<h2>Scale and center the data</h2>
<p>Next, as a standard pre-processing step prior to dimensional reduction techniques like PCA, we center and scale each gene’s expression</p>
<ul>
<li>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate.</li>
</ul>
<p>In addition, the ScaleData function optionally also ‘regresses out’ effects of given confounders, such as the number of RNA molecules or the proportion of mitochrondrial RNA <a href="https://www.biorxiv.org/content/10.1101/576827v2">Hafemeister (2017)</a>. However one should proceed with caution, in case the ‘confounder’ is correlated with biological variables of interest.</p>
<p>The results of this are stored in <code>pbmc@assays$RNA@scale.data</code></p>
<p>Note:</p>
<ul>
<li>Scaling only makes sense as pre-processing for Principal Component Analysis (PCA); for other analysis continue to use LogNormalized data (<code>pbmc@assays$RNA@data</code>) or raw counts (<code>pbmc@assays$RNA@counts</code>)</li>
</ul>
<pre class="r"><code>pbmc &lt;- ScaleData(pbmc, 
                  verbose = T,
                  vars.to.regress = c(&quot;nCount_RNA&quot;, &quot;percent.mt&quot;))</code></pre>
<pre><code>Regressing out nCount_RNA, percent.mt</code></pre>
<pre><code>Scaling data matrix</code></pre>
<pre class="r"><code>pbmc@assays$RNA@scale.data[0:5,0:3]</code></pre>
<pre><code>       AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC
PPBP       -0.1521894    -0.03564544    -0.08450815
LYZ        -0.2679889    -0.82318984     0.08203817
S100A9     -0.8027624    -1.21493665    -0.55611741
IGLL5      -0.1949180    -0.12315643    -0.13727433
GNLY       -0.4260540    -0.21637331     0.85483132</code></pre>
<ul>
<li><em>Contrary to the documentation</em> (see <code>?ScaleData</code>), the function by default scales only highly variable genes</li>
</ul>
<pre class="r"><code>dim(pbmc@assays$RNA@scale.data)</code></pre>
<pre><code>[1] 2000 2638</code></pre>
<p>If you wish to use all genes for PCA, pass the argument <code>features = rownames(pbmc)</code></p>
</div>
</div>
<div id="dimensionality-reduction-and-clustering" class="section level1">
<h1>Dimensionality reduction and clustering</h1>
<div id="perform-linear-dimensional-reduction" class="section level2">
<h2>Perform linear dimensional reduction</h2>
<p>Next we perform Principal Component Analysis (PCA) on the scaled data.</p>
<p>By default, only the highly variable features are used to compute PCA, but can be defined using <code>features</code> argument if you wish to choose a different subset (such as all genes), providing that you have run <code>ScaleData</code> on these genes.</p>
<ul>
<li>Since the PCA algorithm uses a pseudorandom algorithm, provide the <code>seed.use</code> argument for reproducibility.</li>
</ul>
<pre class="r"><code>pbmc &lt;- RunPCA(pbmc, 
               npcs=30, #number of principal components
               features = VariableFeatures(object = pbmc), 
               verbose=F,
               #ndims.print = 1:3, # number of PCs to print
               #nfeatures.print = 10, # number of highly loading genes to print
               seed.use=randomSeed)</code></pre>
<pre class="r"><code>DimPlot(pbmc, reduction = &quot;pca&quot;)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-18-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-18-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="determine-the-real-dimensionality-of-the-dataset" class="section level2">
<h2>Determine the ‘real dimensionality’ of the dataset</h2>
<p>Seurat clusters cells based on their PCA scores. How many components should we use? 10? 20? 50?</p>
<p>Seurat provides three approaches to selecting the number of PCs:</p>
<ol style="list-style-type: decimal">
<li>Select PCs based on whether interesting genes score highly on them (supervised)</li>
<li><code>JackStraw</code>: use a statistical test based on a random null model to determine which PCs capture ‘real’ variation versus noise. Time-consuming but unsupervised and rigourous.</li>
<li><code>ElbowPlot</code>: Use an elbbow plot to select PCs that explain most of the variance in the dataset. A fast heuristic that is commonly used.</li>
</ol>
<div id="select-pcs-based-on-interesting-genes" class="section level3">
<h3>Select PCs based on interesting genes</h3>
<p>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including <code>VizDimLoadings</code>, <code>VizDimReduction</code>, <code>DimPlot</code>, and <code>DimHeatmap</code></p>
<pre class="r"><code>VizDimLoadings(pbmc, dims = 1:2, reduction = &quot;pca&quot;)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-19-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-19-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><code>DimHeatmap</code> plots expression levels in cells and features, ordered according to their PCA scores. The function plots the ‘extreme’ cells on both ends of the spectrum.</p>
<ul>
<li>Note that principal components are just coordinate axes with an arbitrary sign. Hence the sign of a gene or cell ‘score’ on a component does not indicate whether or not the gene is highly or lowly expressed, which is why to look at both the highest and the lowest scoring cells and genes.</li>
</ul>
<pre class="r"><code>DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-20-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-20-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="jackstraw" class="section level3">
<h3>JackStraw</h3>
<p>In <a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">Macosko et al</a> the authors implemented a resampling test inspired by <a href="https://www.ncbi.nlm.nih.gov/pubmed/25336500">Chung et al, Bioinformatics (2014)</a></p>
<p>For each gene</p>
<ol style="list-style-type: decimal">
<li>Randomly ‘scramble’ the gene’s scores across cells and calculate projected PCA loadings for these ‘scambled’ genes. Do this many (200+) times.</li>
<li>Compare the PCA scores for the ‘random’ genes with the observed PCA scores to obtain a p-value for each gene’s association with each principal component.</li>
</ol>
<pre class="r"><code># NOT RUN
pbmc &lt;- JackStraw(pbmc, 
                  num.replicate = 100)</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Use the p-values for each gene per PC to perform a proportion test comparing the number of features with a p-value below a particular threshold (score.thresh), compared with the proportion of features expected under a uniform distribution of p-values. This gives a p-value for each principal component.</li>
</ol>
<pre class="r"><code>pbmc &lt;- ScoreJackStraw(pbmc, 
                       dims = 1:20)</code></pre>
<p>The <code>JackStrawPlot</code> function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.</p>
<pre class="r"><code>JackStrawPlot(pbmc, dims = 1:15)</code></pre>
<pre><code>Warning: Removed 23369 rows containing missing values (geom_point).</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-23-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-23-1">
Past versions of unnamed-chunk-23-1.png
</button>
</p>
<div id="fig-unnamed-chunk-23-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-23-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="elbowplot" class="section level3">
<h3>ElbowPlot</h3>
<p>‘Elbow plot’: shows the percentage of variance explained by each PC. In this example, we can observe an ‘elbow’ around PCs 9-10, suggesting that the majority of true signal is captured in the first 10 PCs.</p>
<pre class="r"><code>ElbowPlot(pbmc)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-24-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-24-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-12 as a cutoff.</p>
<p>We chose 10 here, but consider the following:</p>
<ul>
<li>The composition of the dataset has a large impact on the principal components.
<ul>
<li>Insufficient filtering of cells with unexpectedly high mitochrondrial RNA (indicating cell death) or ribosomal RNA (in some settings indicating stress) may introduce a lot of ‘uninteresting’ variation that affects downstream results</li>
<li>Rare celltypes with few cells will have low weight in the PCA. It is therefore useful to check the gene scores for markers.</li>
<li>If unsure, err on the side of more PCs</li>
</ul></li>
</ul>
</div>
</div>
<div id="cluster-the-cells" class="section level2">
<h2>Cluster the cells</h2>
<p>Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (<a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">Macosko et al</a>). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data <a href="http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract">SNN-Cliq, Xu and Su, Bioinformatics, 2015</a> and CyTOF data <a href="http://www.ncbi.nlm.nih.gov/pubmed/26095251">PhenoGraph, Levine et al., Cell, 2015</a>. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.</p>
<p>As in <code>PhenoGraph</code>, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the <code>FindNeighbors</code> function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).</p>
<p>To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM <a href="http://dx.doi.org/10.1088/1742-5468/2008/10/P10008">SLM, Blondel et al., Journal of Statistical Mechanics</a>, to iteratively group cells together, with the goal of optimizing the standard modularity function.</p>
<p>The <code>FindClusters</code> function, which implements this procedure, contains a <code>resolution</code> parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters.</p>
<ul>
<li>Setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets.</li>
</ul>
<pre class="r"><code>pbmc &lt;- FindNeighbors(pbmc, 
                      dims = 1:10) # use 10 Principal Components</code></pre>
<pre><code>Computing nearest neighbor graph</code></pre>
<pre><code>Computing SNN</code></pre>
<pre class="r"><code>pbmc &lt;- FindClusters(pbmc,
                     resolution = 0.5) </code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2638
Number of edges: 157834

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8590
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
<pre class="r"><code># Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)</code></pre>
<pre><code>AAACATACAACCAC AAACATTGAGCTAC AAACATTGATCAGC AAACCGTGCTTCCG AAACCGTGTATGCG 
             0              2              3              1              6 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
</div>
<div id="run-non-linear-dimensional-reduction-umaptsne" class="section level2">
<h2>Run non-linear dimensional reduction (UMAP/tSNE)</h2>
<p>Seurat offers several non-linear dimensional reduction techniques to visualize and explore these datasets, including</p>
<ul>
<li>t-distributed Stochastic Neighbour Embedding (t-SNE)</li>
<li>Uniform Manifold Approximation and Projection (UMAP)</li>
</ul>
<p>As input to the UMAP and tSNE, use the same PCs as used for the clustering analysis (here, the first 10)</p>
<p>For tSNE, the <code>perplexity</code> parameter (defaults to 30) has a large impact on the final plot. For a great discussion see <a href="https://distill.pub/2016/misread-tsne/">Watternberg et al, 2016</a></p>
<p>As with RunPCA, specifying seed.use makes the reduction reproducible</p>
<pre class="r"><code># use tSNE because UMAP isn&#39;t installed :)
pbmc &lt;- RunTSNE(pbmc, 
                dims = 1:10,
                seed.use=randomSeed,
                perplexity=30, 
                check_duplicates=F) # if not specified the function fails if two cells share coordinates</code></pre>
<pre class="r"><code># note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, 
        reduction = &quot;tsne&quot;)</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-29-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/1d5ea1fcda8330d20343ca2222451609a95f5508/docs/figure/analysis.Rmd/unnamed-chunk-29-1.png" target="_blank">1d5ea1f</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-10
</td>
</tr>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-29-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="biological-analysis" class="section level1">
<h1>Biological analysis</h1>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level2">
<h2>Finding differentially expressed features (cluster biomarkers)</h2>
<p>Seurat can help you find markers that define clusters via differential expression.</p>
<p>FindAllMarkers automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.</p>
<p>Seurat has several tests for differential expression which can be set with the test.use parameter (see our <a href="http://satijalab01.nygenome.org/seurat/v3.0/de_vignette.html">DE vignette</a> for details).</p>
<p>Note that:</p>
<ul>
<li>By default, unless you specify otherwise <code>FindClusters</code> identifies positive <em>and negative markers</em> of a single cluster (specified in ident.1), compared to all other cells</li>
<li>The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells</li>
<li>The thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups.</li>
<li>max.cells.per.ident will downsample each identity class, which can save much time</li>
</ul>
<pre class="r"><code># find all markers of cluster 1
cluster1.markers &lt;- FindMarkers(pbmc, 
                                ident.1 = 1, 
                                test.use  =&quot;wilcox&quot;,
                                only.pos = T,
                                max.cells.per.ident=250,
                                random.seed=randomSeed,
                                min.pct = 0.25)
head(cluster1.markers, n = 5)</code></pre>
<pre><code>              p_val avg_logFC pct.1 pct.2    p_val_adj
S100A9 4.998348e-86  3.854759 0.996 0.217 6.854735e-82
S100A8 4.256255e-85  3.796943 0.973 0.124 5.837028e-81
LYZ    1.681689e-81  3.143552 1.000 0.518 2.306268e-77
LGALS2 6.584607e-76  2.634321 0.907 0.061 9.030130e-72
FCN1   2.484055e-72  2.335357 0.952 0.153 3.406633e-68</code></pre>
<pre class="r"><code># find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers &lt;- FindMarkers(pbmc, 
                                ident.1 = 5, 
                                ident.2 = c(0, 3), 
                                test.use  =&quot;wilcox&quot;,
                                only.pos = T,
                                max.cells.per.ident=250,
                                random.seed=randomSeed,
                                min.pct = 0.25)
head(cluster5.markers, n = 5)</code></pre>
<pre><code>              p_val avg_logFC pct.1 pct.2    p_val_adj
TYROBP 4.755568e-77  3.239007 1.000 0.133 6.521786e-73
FCER1G 2.124871e-76  3.286471 1.000 0.101 2.914048e-72
FCGR3A 8.797046e-76  2.901285 0.969 0.047 1.206427e-71
LST1   8.597783e-75  3.310554 1.000 0.164 1.179100e-70
AIF1   5.072484e-74  2.851443 1.000 0.228 6.956404e-70</code></pre>
<pre class="r"><code># find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers &lt;- FindAllMarkers(pbmc, 
                               test.use=&quot;wilcox&quot;,
                               only.pos = TRUE, 
                               min.pct = 0.25, 
                               max.cells.per.ident=250,
                               random.seed=randomSeed,
                               logfc.threshold = 0.25)</code></pre>
<pre><code>Calculating cluster 0</code></pre>
<pre><code>Calculating cluster 1</code></pre>
<pre><code>Calculating cluster 2</code></pre>
<pre><code>Calculating cluster 3</code></pre>
<pre><code>Calculating cluster 4</code></pre>
<pre><code>Calculating cluster 5</code></pre>
<pre><code>Calculating cluster 6</code></pre>
<pre><code>Calculating cluster 7</code></pre>
<pre><code>Calculating cluster 8</code></pre>
<pre class="r"><code>pbmc.markers %&gt;% group_by(cluster) %&gt;% top_n(n = 2, wt = avg_logFC)</code></pre>
<pre><code># A tibble: 18 x 7
# Groups:   cluster [9]
      p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene    
      &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;   
 1 1.95e-32     0.857 0.931 0.56   2.67e-28 0       LDHB    
 2 5.26e-20     0.931 0.434 0.089  7.22e-16 0       CCR7    
 3 5.00e-86     3.85  0.996 0.217  6.85e-82 1       S100A9  
 4 4.26e-85     3.80  0.973 0.124  5.84e-81 1       S100A8  
 5 5.61e-83     2.99  0.936 0.042  7.69e-79 2       CD79A   
 6 4.47e-47     2.48  0.62  0.023  6.12e-43 2       TCL1A   
 7 1.28e-27     0.891 0.947 0.494  1.75e-23 3       IL32    
 8 4.69e-11     0.911 0.379 0.136  6.43e- 7 3       AQP3    
 9 3.37e-69     2.14  0.961 0.232  4.62e-65 4       CCL5    
10 6.18e-37     2.06  0.588 0.051  8.47e-33 4       GZMK    
11 4.07e-69     2.28  0.969 0.135  5.59e-65 5       FCGR3A  
12 9.40e-65     2.17  1     0.314  1.29e-60 5       LST1    
13 5.76e-75     3.39  0.986 0.07   7.90e-71 6       GZMB    
14 3.42e-65     3.47  0.959 0.134  4.69e-61 6       GNLY    
15 4.20e-46     2.69  0.806 0.012  5.76e-42 7       FCER1A  
16 1.26e-17     1.99  1     0.513  1.73e-13 7       HLA-DPB1
17 1.68e-49     5.94  1     0.024  2.30e-45 8       PPBP    
18 1.68e-49     5.02  1     0.01   2.30e-45 8       PF4     </code></pre>
<p>Seurat includes several tools for visualizing marker expression.</p>
<ul>
<li><code>VlnPlot</code> shows expression probability distributions across clusters</li>
<li><code>FeaturePlot</code> visualizes feature expression on a tSNE or PCA plot</li>
<li><code>RidgePlot</code> draws a ridge plot of gene expression, metrics, PC scores, etc.</li>
<li><code>CellScatter</code> creates a plot of scatter plot of features across two single cells</li>
<li><code>DotPlot</code> shows average gene expression across different identity classes</li>
</ul>
<pre class="r"><code>VlnPlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;CD79A&quot;))</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-33-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-33-1">
Past versions of unnamed-chunk-33-1.png
</button>
</p>
<div id="fig-unnamed-chunk-33-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-33-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>FeaturePlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;GNLY&quot;, &quot;CD3E&quot;, &quot;CD14&quot;, &quot;FCER1A&quot;, &quot;FCGR3A&quot;, &quot;LYZ&quot;, &quot;PPBP&quot;, &quot;CD8A&quot;))</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-34-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-34-1">
Past versions of unnamed-chunk-34-1.png
</button>
</p>
<div id="fig-unnamed-chunk-34-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-34-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="assigning-cell-type-identity-to-clusters" class="section level2">
<h2>Assigning cell type identity to clusters</h2>
<p>With this dataset, we cheat a bit by using canonical markers to easily match the differentially expressed genes of unbiased cell clusters to known cell types. (An alternative might have been to align the cells against an existing labelled dataset and transfer its labels)</p>
<pre class="r"><code>new.cluster.ids &lt;- c(&quot;Naive CD4 T&quot;, &quot;Memory CD4 T&quot;, &quot;CD14+ Mono&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;FCGR3A+ Mono&quot;, 
    &quot;NK&quot;, &quot;DC&quot;, &quot;Platelet&quot;)
names(new.cluster.ids) &lt;- levels(pbmc)
pbmc &lt;- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = &quot;tsne&quot;, label = TRUE, label.size = 5, pt.size = 0.5) + NoLegend()</code></pre>
<p><img src="figure/analysis.Rmd/unnamed-chunk-35-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-35-1">
Past versions of unnamed-chunk-35-1.png
</button>
</p>
<div id="fig-unnamed-chunk-35-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/JonThom/190510-scWorkshop/blob/fcf3e9123afdba50085b224644448da495963e5f/docs/figure/analysis.Rmd/unnamed-chunk-35-1.png" target="_blank">fcf3e91</a>
</td>
<td>
JonThom
</td>
<td>
2019-05-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="wrap-up" class="section level1">
<h1>Wrap up</h1>
<p>Save the final seurat object to disk</p>
<pre class="r"><code>if (F) saveRDS(pbmc, file = &quot;../output/pbmc3k_final.rds&quot;)</code></pre>
<div id="additional-material" class="section level3">
<h3>Additional material</h3>
<p><a href="https://satijalab.org/seurat/">Satija group website</a> - for more tutorials and articles</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Storage

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.3.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.8.0.1     Seurat_3.0.0.9000

loaded via a namespace (and not attached):
 [1] httr_1.3.1         tidyr_0.8.2        viridisLite_0.3.0 
 [4] jsonlite_1.5       splines_3.5.3      R.utils_2.6.0     
 [7] gtools_3.8.1       assertthat_0.2.1   yaml_2.1.19       
[10] ggrepel_0.8.0      globals_0.12.4     pillar_1.3.1      
[13] backports_1.1.2    lattice_0.20-38    reticulate_1.8    
[16] glue_1.3.1         digest_0.6.18      RColorBrewer_1.1-2
[19] SDMTools_1.1-221   colorspace_1.4-1   cowplot_0.9.2     
[22] htmltools_0.3.6    Matrix_1.2-15      R.oo_1.22.0       
[25] plyr_1.8.4         pkgconfig_2.0.2    tsne_0.1-3        
[28] listenv_0.7.0      purrr_0.3.0        scales_1.0.0      
[31] RANN_2.5.1         gdata_2.18.0       whisker_0.3-2     
[34] Rtsne_0.13         git2r_0.23.0       tibble_2.1.1      
[37] ggplot2_3.1.0      withr_2.1.2        ROCR_1.0-7        
[40] pbapply_1.3-4      lazyeval_0.2.2     cli_1.1.0         
[43] survival_2.43-3    magrittr_1.5       crayon_1.3.4      
[46] evaluate_0.10.1    R.methodsS3_1.7.1  fansi_0.4.0       
[49] fs_1.2.6           future_1.10.0      nlme_3.1-137      
[52] MASS_7.3-51.1      gplots_3.0.1.1     ica_1.0-2         
[55] tools_3.5.3        fitdistrplus_1.0-9 data.table_1.12.2 
[58] stringr_1.4.0      plotly_4.7.1       munsell_0.5.0     
[61] cluster_2.0.7-1    irlba_2.3.2        compiler_3.5.3    
[64] rsvd_1.0.0         caTools_1.17.1.1   rlang_0.3.3       
[67] grid_3.5.3         ggridges_0.5.0     htmlwidgets_1.2   
[70] igraph_1.2.1       labeling_0.3       bitops_1.0-6      
[73] rmarkdown_1.10     gtable_0.3.0       codetools_0.2-16  
[76] R6_2.4.0           zoo_1.8-2          knitr_1.20        
[79] utf8_1.1.4         future.apply_1.0.1 workflowr_1.3.0   
[82] rprojroot_1.3-2    KernSmooth_2.23-15 metap_0.9         
[85] ape_5.1            stringi_1.4.3      parallel_3.5.3    
[88] Rcpp_1.0.1         png_0.1-7          tidyselect_0.2.5  
[91] lmtest_0.9-36     </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
